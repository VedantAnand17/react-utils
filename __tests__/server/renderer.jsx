import _ from 'lodash';
import factory, { SCRIPT_LOCATIONS } from 'server/renderer';
import fs from 'fs';
import React from 'react';

import { Helmet } from 'react-helmet';
import { Route } from 'react-router-dom';
import { createStore } from 'redux';

let mockFailsForgeRandomGetBytesMethod = false;

jest.mock('node-forge', () => {
  const mockForge = require.requireActual('node-forge');
  mockForge.random.getBytes = (numBytes, cb) => {
    if (mockFailsForgeRandomGetBytesMethod) {
      cb(new Error('Emulated Failure of forge.random.getBytes(..)'));
    }
    let res = '';
    for (let i = 0; i < numBytes; i += 1) res += i % 10;
    cb(null, res);
  };
  return mockForge;
});

const TEST_CONTEXT = `${__dirname}/test_data`;

const TEST_REDUX_STATE = {
  stateKey1: 'State Value #1',
  stateKey2: 'State Value #2',
};

const TEST_HTTP_REQUEST = {
  info: 'I am a dummy HTTP request! No need for a complex mock here!',
};

const TEST_WEBPACK_CONFIG = {
  context: TEST_CONTEXT,
  output: {
    path: '/test/path',
    publicPath: '/test/public/path/',
  },
};

const testBuildInfo = JSON.parse(
  fs.readFileSync(`${TEST_CONTEXT}/.build-info`),
);

beforeAll(() => {
  Helmet.canUseDOM = false;
});

afterEach(() => {
  delete global.TRU_BUILD_INFO;
  mockFailsForgeRandomGetBytesMethod = false;
});

afterAll(() => {
  Helmet.canUseDOM = true;
});

/**
 * Performs tests with the specified configuration and options.
 * @param {Object} webpackConfig
 * @param {Object} options
 * @return {Promise}
 */
async function coreTest(webpackConfig, options) {
  expect(global.TRU_BUILD_INFO).toBeUndefined();

  let renderer;
  expect(() => {
    renderer = factory(webpackConfig, options);
  }).not.toThrow();
  expect(renderer).toBeInstanceOf(Function);
  expect(global.TRU_BUILD_INFO).toEqual(testBuildInfo);

  try {
    let render = '';
    await renderer(_.clone(TEST_HTTP_REQUEST), {
      locals: {
        webpackStats: {
          toJson: () => ({
            assetsByChunkName: {
              'test-chunk-a': 'test-chunk-a-1511941200000.css',
              'test-chunk-b': [
                'test-chunk-b-1511941200000.js',
                'test-chunk-b-1511941200000.css',
              ],
            },
          }),
        },
      },
      send: (x) => { render += x; },
      status: (x) => { render += `HTTP STATUS: ${x}\n`; },
    });
    expect(render).toMatchSnapshot();
  } catch (e) {
    expect(e).toMatchSnapshot();
  }
}

test('Base rendering of HTML template',
  () => coreTest(TEST_WEBPACK_CONFIG, {}));

test('Config overriding for injection',
  () => coreTest(TEST_WEBPACK_CONFIG, {
    beforeRender: async (res, sanitizedConfig) => {
      expect(res).toEqual(TEST_HTTP_REQUEST);
      expect(sanitizedConfig).toBeInstanceOf(Object);
      expect(sanitizedConfig).not.toHaveProperty('SECRET');
      return {
        configToInject: {
          ...sanitizedConfig,
          additionalKey: 'Additional Value',
        },
      };
    },
  }));

test('Hemlet integration works',
  () => coreTest(TEST_WEBPACK_CONFIG, {
    Application: () => (
      <div>
        <p>
          Hello World!
        </p>
        <Helmet>
          <title>
            Test Page Title
          </title>
          <meta property="description" content="Test Page Description" />
        </Helmet>
        <p>
          Goodbye World!
        </p>
      </div>
    ),
  }));

test('Injection of additional JS scripts',
  () => coreTest(TEST_WEBPACK_CONFIG, {
    beforeRender: async () => ({
      extraScripts: [
        '<script>Dummy JS Sript</script>',
        '<script>Another Dummy JS Script</script>', {
          code: '<script>Yet another Dummy JS Script</script>',
          location: SCRIPT_LOCATIONS.DEFAULT,
        }, {
          code: '<script>1-st script after opening <head></script>',
          location: SCRIPT_LOCATIONS.HEAD_OPEN,
        }, {
          code: '<script>1-st script after opening <body></script>',
          location: SCRIPT_LOCATIONS.BODY_OPEN,
        }, {
          code: '<script>2-nd script after opening <body></script>',
          location: SCRIPT_LOCATIONS.BODY_OPEN,
        }, {
          code: '<script>2-nd script after opening <head></script>',
          location: SCRIPT_LOCATIONS.HEAD_OPEN,
        },
      ],
    }),
  }));

test('Server-side rendering (SSR); injection of CSS chunks & Redux state',
  () => coreTest(TEST_WEBPACK_CONFIG, {
    Application: () => (
      <Route
        component={({ staticContext }) => {
          staticContext.chunks.push('test-chunk-a');
          staticContext.chunks.push('test-chunk-b');
          return (
            <div>
              Hello Wold!
            </div>
          );
        }}
      />
    ),
    beforeRender: async () => ({
      store: createStore(() => _.clone(TEST_REDUX_STATE)),
    }),
  }));

test('Setting of response HTTP status the server-side rendering', () => {
  coreTest(TEST_WEBPACK_CONFIG, {
    Application: () => (
      <Route
        component={({ staticContext }) => {
          staticContext.status = 404; // eslint-disable-line no-param-reassign
          return (
            <div>
              404 Error Test
            </div>
          );
        }}
      />
    ),
  });
});

test('Throws in case of forge.random.getBytes(..) failure', () => {
  mockFailsForgeRandomGetBytesMethod = true;
  return coreTest(TEST_WEBPACK_CONFIG, {});
});
