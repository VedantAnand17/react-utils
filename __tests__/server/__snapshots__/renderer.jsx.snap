// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`"favicon" option 1`] = `
"<!DOCTYPE html>
        <html lang=\\"en\\">
          <head>
            
            
            
            <meta name=\\"theme-color\\" content=\\"#FFFFFF\\"/>
            
            <link href=\\"/test/public/path/main-1511941200000.css\\" rel=\\"stylesheet\\" />
            <link rel=\\"shortcut icon\\" href=\\"/favicon.ico\\" />
            <meta charset=\\"utf-8\\" />
            <meta
              content=\\"width=device-width,initial-scale=1.0\\"
              name=\\"viewport\\"
            />
          </head>
          <body>
            
            <div id=\\"react-view\\"></div>
            <script
              id=\\"inj\\"
              type=\\"application/javascript\\"
              nonce=\\"abcdef-dummy-nonce\\"
            >
              window.SPLITS = {}
              window.INJ=\\"MDEyMzQ1Njc4OTAxMjM0NTY3ODkwMTIzNDU2Nzg5MDGxqHiEs7kHDY44Xxn5w1REclHuC2M3QGvQLK2naeWTzCJpXWctF+ki6HirgIkIyx4BfBIqlJbTLnShL1wyMPDARKNI4xpk9YGnthf9zGmb1ZElRTRh7QU4i6pTuK68/j92srOJYT+ccoyqPwGMJheURmW3JxPMKpqVhILhX7+HB82hJ5hxYefuQi4Y2Ad+zmnB1q/YWY+3O8tfpLoprbCH8lG9bQF8y4L3QXepFHWKkGQrxc6kLWmC9u7yEbUehrEnpZXFcp1Okw8Ox0gxT6GYtbpbMKeQmydOpe4ag6ReKjpokG/1ivpFHdgH+eVbB+sESBiVozhNa9vPJ6FBHKN2GDlyf5Gy7oUkFexI6ozoxxhBmdxFcm8+BvO9nuC7cThbhLgot65FgVWXtWnkhh5xA+viJ5vnfnY2/UbytYsQvQ==\\"
            </script>
            <script src=\\"/test/public/path/polyfills-1511941200000.js\\" type=\\"application/javascript\\"></script><script src=\\"/test/public/path/main-1511941200000.js\\" type=\\"application/javascript\\"></script>
            
          </body>
        </html>"
`;

exports[`Base rendering of HTML template 1`] = `
"<!DOCTYPE html>
        <html lang=\\"en\\">
          <head>
            
            
            
            <meta name=\\"theme-color\\" content=\\"#FFFFFF\\"/>
            
            <link href=\\"/test/public/path/main-1511941200000.css\\" rel=\\"stylesheet\\" />
            
            <meta charset=\\"utf-8\\" />
            <meta
              content=\\"width=device-width,initial-scale=1.0\\"
              name=\\"viewport\\"
            />
          </head>
          <body>
            
            <div id=\\"react-view\\"></div>
            <script
              id=\\"inj\\"
              type=\\"application/javascript\\"
              nonce=\\"abcdef-dummy-nonce\\"
            >
              window.SPLITS = {}
              window.INJ=\\"MDEyMzQ1Njc4OTAxMjM0NTY3ODkwMTIzNDU2Nzg5MDGxqHiEs7kHDY44Xxn5w1REclHuC2M3QGvQLK2naeWTzCJpXWctF+ki6HirgIkIyx4BfBIqlJbTLnShL1wyMPDARKNI4xpk9YGnthf9zGmb1ZElRTRh7QU4i6pTuK68/j92srOJYT+ccoyqPwGMJheURmW3JxPMKpqVhILhX7+HB82hJ5hxYefuQi4Y2Ad+zmnB1q/YWY+3O8tfpLoprbCH8lG9bQF8y4L3QXepFHWKkGQrxc6kLWmC9u7yEbUehrEnpZXFcp1Okw8Ox0gxT6GYtbpbMKeQmydOpe4ag6ReKjpokG/1ivpFHdgH+eVbB+sESBiVozhNa9vPJ6FBHKN2GDlyf5Gy7oUkFexI6ozoxxhBmdxFcm8+BvO9nuC7cThbhLgot65FgVWXtWnkhh5xA+viJ5vnfnY2/UbytYsQvQ==\\"
            </script>
            <script src=\\"/test/public/path/polyfills-1511941200000.js\\" type=\\"application/javascript\\"></script><script src=\\"/test/public/path/main-1511941200000.js\\" type=\\"application/javascript\\"></script>
            
          </body>
        </html>"
`;

exports[`Config overriding for injection 1`] = `
"<!DOCTYPE html>
        <html lang=\\"en\\">
          <head>
            
            
            
            <meta name=\\"theme-color\\" content=\\"#FFFFFF\\"/>
            
            <link href=\\"/test/public/path/main-1511941200000.css\\" rel=\\"stylesheet\\" />
            
            <meta charset=\\"utf-8\\" />
            <meta
              content=\\"width=device-width,initial-scale=1.0\\"
              name=\\"viewport\\"
            />
          </head>
          <body>
            
            <div id=\\"react-view\\"></div>
            <script
              id=\\"inj\\"
              type=\\"application/javascript\\"
              nonce=\\"abcdef-dummy-nonce\\"
            >
              window.SPLITS = {}
              window.INJ=\\"MDEyMzQ1Njc4OTAxMjM0NTY3ODkwMTIzNDU2Nzg5MDGxqHiEs7kHDY44Xxn5w1REclHuC2M3QGvQLK2naeWTzCJpXWctF+ki6HirgIkIyx4BfBIqlJbTLnShL1wyMPDARKNI4xpk9YGnthf9zGmb1ZElRTRh7QU4i6pTuK68/j92srOJYT+ccoyqPwGMJheURmW3JxPMKpqVhILhX7+HB82hJ5hxYefuQi4Y2Ad+zmnB1q/YWY+3O8tfpLoprbCH8lG9bQF8y4L3QXepFHWKkGQrxc6kLWmC9u7yEbUehrEnpZXFcp1Okw8Ox0gxT6GYtbpbMKeQmydOpe4ag6ReKjpokG/1ivpFHdgH+eVbB+sESBiVozhNa9vPJ6FBHKN2GDlyf5Gy7oUkFexI6ozoxxhBmdxFcm8+BvO9nuC7cTi9VQX8OW0+JoVNYcuVavL1zUfXa/F+j/uVT8wF5mV+OD+JFE4PXHgKEkoOxVKClTYCuusupNawOFnXjAKzQTAr1w2Y1xfsnhT1PDqLQP3ngQ==\\"
            </script>
            <script src=\\"/test/public/path/polyfills-1511941200000.js\\" type=\\"application/javascript\\"></script><script src=\\"/test/public/path/main-1511941200000.js\\" type=\\"application/javascript\\"></script>
            
          </body>
        </html>"
`;

exports[`Hemlet integration works 1`] = `
"<!DOCTYPE html>
        <html lang=\\"en\\">
          <head>
            
            <title data-react-helmet=\\"true\\">Test Page Title</title>
            <meta data-react-helmet=\\"true\\" property=\\"description\\" content=\\"Test Page Description\\"/>
            <meta name=\\"theme-color\\" content=\\"#FFFFFF\\"/>
            
            <link href=\\"/test/public/path/main-1511941200000.css\\" rel=\\"stylesheet\\" />
            
            <meta charset=\\"utf-8\\" />
            <meta
              content=\\"width=device-width,initial-scale=1.0\\"
              name=\\"viewport\\"
            />
          </head>
          <body>
            
            <div id=\\"react-view\\"><div><p>Hello World!</p><p>Goodbye World!</p></div></div>
            <script
              id=\\"inj\\"
              type=\\"application/javascript\\"
              nonce=\\"abcdef-dummy-nonce\\"
            >
              window.SPLITS = {}
              window.INJ=\\"MDEyMzQ1Njc4OTAxMjM0NTY3ODkwMTIzNDU2Nzg5MDGxqHiEs7kHDY44Xxn5w1REclHuC2M3QGvQLK2naeWTzCJpXWctF+ki6HirgIkIyx4BfBIqlJbTLnShL1wyMPDARKNI4xpk9YGnthf9zGmb1ZElRTRh7QU4i6pTuK68/j92srOJYT+ccoyqPwGMJheURmW3JxPMKpqVhILhX7+HB82hJ5hxYefuQi4Y2Ad+zmnB1q/YWY+3O8tfpLoprbCH8lG9bQF8y4L3QXepFHWKkGQrxc6kLWmC9u7yEbUehrEnpZXFcp1Okw8Ox0gxT6GYtbpbMKeQmydOpe4ag6ReKjpokG/1ivpFHdgH+eVbB+sESBiVozhNa9vPJ6FBHKN2GDlyf5Gy7oUkFexI6ozoxxhBmdxFcm8+BvO9nuC7cThbhLgot65FgVWXtWnkhh5xA+viJ5vnfnY2/UbytYsQvQ==\\"
            </script>
            <script src=\\"/test/public/path/polyfills-1511941200000.js\\" type=\\"application/javascript\\"></script><script src=\\"/test/public/path/main-1511941200000.js\\" type=\\"application/javascript\\"></script>
            
          </body>
        </html>"
`;

exports[`Injection of additional JS scripts 1`] = `
"<!DOCTYPE html>
        <html lang=\\"en\\">
          <head>
            <script>1-st script after opening <head></script><script>2-nd script after opening <head></script>
            
            
            <meta name=\\"theme-color\\" content=\\"#FFFFFF\\"/>
            
            <link href=\\"/test/public/path/main-1511941200000.css\\" rel=\\"stylesheet\\" />
            
            <meta charset=\\"utf-8\\" />
            <meta
              content=\\"width=device-width,initial-scale=1.0\\"
              name=\\"viewport\\"
            />
          </head>
          <body>
            <script>1-st script after opening <body></script><script>2-nd script after opening <body></script>
            <div id=\\"react-view\\"></div>
            <script
              id=\\"inj\\"
              type=\\"application/javascript\\"
              nonce=\\"abcdef-dummy-nonce\\"
            >
              window.SPLITS = {}
              window.INJ=\\"MDEyMzQ1Njc4OTAxMjM0NTY3ODkwMTIzNDU2Nzg5MDGxqHiEs7kHDY44Xxn5w1REclHuC2M3QGvQLK2naeWTzCJpXWctF+ki6HirgIkIyx4BfBIqlJbTLnShL1wyMPDARKNI4xpk9YGnthf9zGmb1ZElRTRh7QU4i6pTuK68/j92srOJYT+ccoyqPwGMJheURmW3JxPMKpqVhILhX7+HB82hJ5hxYefuQi4Y2Ad+zmnB1q/YWY+3O8tfpLoprbCH8lG9bQF8y4L3QXepFHWKkGQrxc6kLWmC9u7yEbUehrEnpZXFcp1Okw8Ox0gxT6GYtbpbMKeQmydOpe4ag6ReKjpokG/1ivpFHdgH+eVbB+sESBiVozhNa9vPJ6FBHKN2GDlyf5Gy7oUkFexI6ozoxxhBmdxFcm8+BvO9nuC7cThbhLgot65FgVWXtWnkhh5xA+viJ5vnfnY2/UbytYsQvQ==\\"
            </script>
            <script src=\\"/test/public/path/polyfills-1511941200000.js\\" type=\\"application/javascript\\"></script><script src=\\"/test/public/path/main-1511941200000.js\\" type=\\"application/javascript\\"></script>
            <script>Dummy JS Sript</script><script>Another Dummy JS Script</script><script>Yet another Dummy JS Script</script>
          </body>
        </html>"
`;

exports[`Server-side rendering (SSR); injection of CSS chunks & Redux state 1`] = `
"<!DOCTYPE html>
        <html lang=\\"en\\">
          <head>
            
            <title data-react-helmet=\\"true\\"></title>
            
            <meta name=\\"theme-color\\" content=\\"#FFFFFF\\"/>
            
            <link href=\\"/test/public/path/main-1511941200000.css\\" rel=\\"stylesheet\\" /><link href=\\"/test/public/path/test-chunk-a-1511941200000.css\\" rel=\\"stylesheet\\" /><link href=\\"/test/public/path/test-chunk-b-1511941200000.css\\" rel=\\"stylesheet\\" />
            
            <meta charset=\\"utf-8\\" />
            <meta
              content=\\"width=device-width,initial-scale=1.0\\"
              name=\\"viewport\\"
            />
          </head>
          <body>
            
            <div id=\\"react-view\\"><div>Hello Wold!</div></div>
            <script
              id=\\"inj\\"
              type=\\"application/javascript\\"
              nonce=\\"abcdef-dummy-nonce\\"
            >
              window.SPLITS = {}
              window.INJ=\\"MDEyMzQ1Njc4OTAxMjM0NTY3ODkwMTIzNDU2Nzg5MDGxqHiEs7kHDY44Xxn5w1REclHuC2M3QGvQLK2naeWTzCJpXWctF+ki6HirgIkIyx4BfBIqlJbTLnShL1wyMPDARKNI4xpk9YGnthf9zGmb1ZElRTRh7QU4i6pTuK68/j92srOJYT+ccoyqPwGMJheURmW3JxPMKpqVhILhX7+HB82hJ5hxYefuQi4Y2Ad+zmnB1q/YWY+3O8tfpLoprbCH8lG9bQF8y4L3QXepFHWKkGQrxc6kLWmC9u7yEbUehrEnpZXFcp1Okw8Ox0gxT6GYtbpbMKeQmydOpe4ag6ReKjpokG/1ivpFHdgH+eVbB+sESBiVozhNa9vPJ6FBHKN2GDlyf5Gy7oUkFexI6ozoxxhBmdxFcm8+BvO9nuC7cThbhLgot65FgVWXtWnkhh5xv8WlAVnlWdgNvB/Q3AKydAbxyFdVF8mjQAb1ExDvse1qjIa7JIOESht5ICAGbdminx9y/KAn4bYuq7OEVZTV+owDk7nfArmDZTvVMMFaL0c=\\"
            </script>
            <script src=\\"/test/public/path/polyfills-1511941200000.js\\" type=\\"application/javascript\\"></script><script src=\\"/test/public/path/main-1511941200000.js\\" type=\\"application/javascript\\"></script><script src=\\"/test/public/path/test-chunk-b-1511941200000.js\\" type=\\"application/javascript\\"></script>
            
          </body>
        </html>"
`;

exports[`Setting of response HTTP status the server-side rendering 1`] = `
"HTTP STATUS: 404
<!DOCTYPE html>
        <html lang=\\"en\\">
          <head>
            
            <title data-react-helmet=\\"true\\"></title>
            
            <meta name=\\"theme-color\\" content=\\"#FFFFFF\\"/>
            
            <link href=\\"/test/public/path/main-1511941200000.css\\" rel=\\"stylesheet\\" />
            
            <meta charset=\\"utf-8\\" />
            <meta
              content=\\"width=device-width,initial-scale=1.0\\"
              name=\\"viewport\\"
            />
          </head>
          <body>
            
            <div id=\\"react-view\\"><div>404 Error Test</div></div>
            <script
              id=\\"inj\\"
              type=\\"application/javascript\\"
              nonce=\\"abcdef-dummy-nonce\\"
            >
              window.SPLITS = {}
              window.INJ=\\"MDEyMzQ1Njc4OTAxMjM0NTY3ODkwMTIzNDU2Nzg5MDGxqHiEs7kHDY44Xxn5w1REclHuC2M3QGvQLK2naeWTzCJpXWctF+ki6HirgIkIyx4BfBIqlJbTLnShL1wyMPDARKNI4xpk9YGnthf9zGmb1ZElRTRh7QU4i6pTuK68/j92srOJYT+ccoyqPwGMJheURmW3JxPMKpqVhILhX7+HB82hJ5hxYefuQi4Y2Ad+zmnB1q/YWY+3O8tfpLoprbCH8lG9bQF8y4L3QXepFHWKkGQrxc6kLWmC9u7yEbUehrEnpZXFcp1Okw8Ox0gxT6GYtbpbMKeQmydOpe4ag6ReKjpokG/1ivpFHdgH+eVbB+sESBiVozhNa9vPJ6FBHKN2GDlyf5Gy7oUkFexI6ozoxxhBmdxFcm8+BvO9nuC7cThbhLgot65FgVWXtWnkhh5xA+viJ5vnfnY2/UbytYsQvQ==\\"
            </script>
            <script src=\\"/test/public/path/polyfills-1511941200000.js\\" type=\\"application/javascript\\"></script><script src=\\"/test/public/path/main-1511941200000.js\\" type=\\"application/javascript\\"></script>
            
          </body>
        </html>"
`;

exports[`Throws in case of forge.random.getBytes(..) failure 1`] = `[Error: Emulated Failure of forge.random.getBytes(..)]`;

exports[`Throws in case of forge.random.getBytes(..) failure 2`] = `""`;
